From d2049d04f4ec07c25f6e7526cb78bf4e40e3611d Mon Sep 17 00:00:00 2001
From: Huang Shijie <b32955@freescale.com>
Date: Wed, 7 May 2014 17:15:03 +0800
Subject: [PATCH 0957/1074] MLK-9811 mtd: gpmi: add supend/resume support

support gpmi nand suspend/resume function

Signed-off-by: Huang Shijie <b32955@freescale.com>
Signed-off-by: Allen Xu <b45815@freescale.com>
(cherry picked from commit d5bad6203191a64ac78bddbf2385da3606b94460)
[Original patch taken from git://git.freescale.com/imx/linux-2.6-imx.git]
Signed-off-by: Biyao Zhai <biyao.zhai@windriver.com>
---
 drivers/mtd/nand/gpmi-nand/gpmi-nand.c |   45 ++++++++++++++++++++++++++++++++
 1 files changed, 45 insertions(+), 0 deletions(-)

diff --git a/drivers/mtd/nand/gpmi-nand/gpmi-nand.c b/drivers/mtd/nand/gpmi-nand/gpmi-nand.c
index f638cd8..15f474e 100644
--- a/drivers/mtd/nand/gpmi-nand/gpmi-nand.c
+++ b/drivers/mtd/nand/gpmi-nand/gpmi-nand.c
@@ -1829,9 +1829,54 @@ static int gpmi_nand_remove(struct platform_device *pdev)
 	return 0;
 }
 
+static int gpmi_pm_suspend(struct device *dev)
+{
+	struct gpmi_nand_data *this = dev_get_drvdata(dev);
+
+	release_dma_channels(this);
+	return 0;
+}
+
+static int gpmi_pm_resume(struct device *dev)
+{
+	struct gpmi_nand_data *this = dev_get_drvdata(dev);
+	int ret;
+
+	if (GPMI_IS_MX6SX(this)) {
+		ret = acquire_dma_channels(this);
+		if (ret < 0)
+			return ret;
+
+		/* re-init the GPMI registers */
+		this->flags &= ~GPMI_TIMING_INIT_OK;
+		ret = gpmi_init(this);
+		if (ret) {
+			dev_err(this->dev, "Error setting GPMI : %d\n", ret);
+			return ret;
+		}
+
+		/* re-init the BCH registers */
+		ret = bch_set_geometry(this);
+		if (ret) {
+			dev_err(this->dev, "Error setting BCH : %d\n", ret);
+			return ret;
+		}
+
+		/* re-init others */
+		gpmi_extra_init(this);
+	}
+
+	return 0;
+}
+
+static const struct dev_pm_ops gpmi_pm_ops = {
+	SET_SYSTEM_SLEEP_PM_OPS(gpmi_pm_suspend, gpmi_pm_resume)
+};
+
 static struct platform_driver gpmi_nand_driver = {
 	.driver = {
 		.name = "gpmi-nand",
+		.pm = &gpmi_pm_ops,
 		.of_match_table = gpmi_nand_id_table,
 	},
 	.probe   = gpmi_nand_probe,
-- 
1.7.5.4

